plugins/inmemory-records-storage/src/main/java/com/github/bsideup/liiklus/records/inmemory/InMemoryRecordsStorage.java

```diff
@@ -0,0 +1,207 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxProcessor;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.ReplayProcessor;
+
+import java.nio.ByteBuffer;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * WARNING: this storage type should only be used for testing and NOT in production
+ */
+@RequiredArgsConstructor
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+public class InMemoryRecordsStorage implements RecordsStorage {
+
+    public static int partitionByKey(String key, int numberOfPartitions) {
+        return partitionByKey(ByteBuffer.wrap(key.getBytes()), numberOfPartitions);
+    }
+
+    public static int partitionByKey(ByteBuffer key, int numberOfPartitions) {
+        return Math.abs(key.hashCode()) % numberOfPartitions;
+    }
+
+    int numberOfPartitions;
+
+    ConcurrentMap<String, StoredTopic> state = new ConcurrentHashMap<>();
+
+    @Override
+    public CompletionStage<OffsetInfo> publish(Envelope envelope) {
+        val topic = envelope.getTopic();
+        val storedTopic = state.computeIfAbsent(topic, __ -> new StoredTopic(numberOfPartitions));
+
+        val partition = partitionByKey(envelope.getKey(), numberOfPartitions);
+        val storedPartition = storedTopic.getPartitions().computeIfAbsent(
+                partition,
+                __ -> new StoredTopic.StoredPartition()
+        );
+
+        val offset = storedPartition.getNextOffset().getAndIncrement();
+        storedPartition.getProcessor().onNext(new StoredTopic.StoredPartition.StoredRecord(
+                offset,
+                envelope.getKey(),
+                envelope.getValue()
+        ));
+
+        return CompletableFuture.completedFuture(new OffsetInfo(
+                topic,
+                partition,
+                offset
+        ));
+    }
+
+    @Override
+    public Subscription subscribe(String topic, String groupName, Optional<String> autoOffsetReset) {
+        val storedTopic = state.computeIfAbsent(topic, __ -> new StoredTopic(numberOfPartitions));
+        return new Subscription() {
+
+            @Override
+            public Publisher<Stream<? extends PartitionSource>> getPublisher(
+                    Supplier<CompletionStage<Map<Integer, Long>>> offsetsProvider
+            ) {
+                val subscription = this;
+                return Flux.create(sink -> {
+
+                    sink.onCancel(() -> storedTopic.revoke(groupName, this));
+
+                    if (sink.isCancelled()) {
```

Comment:
should it be before onCancel callback setup?

Reply:
AFAIK it shouldn't. We set `onCancel` before everything else, then we do the main logic unless it's already canceled. Otherwise there will be a race condition